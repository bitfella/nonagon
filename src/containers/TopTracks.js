import React, { Component } from 'react';
import { connect } from 'react-redux';
import { isEmpty } from 'lodash-es';
import styled from 'styled-components/macro';
import PropTypes from 'prop-types';
import handleError from 'services/handleError';
import { getTopTracks } from 'actions/tastes';
import { appTimeRange } from 'config/spotify';
import {
  getToken,
  getTastesTracks,
  tastesLoading,
  tastesError
} from 'selectors';
import Track from 'components/Track';
import TopTracksActions from 'components/TopTracksActions';
import PlaceholderList from 'components/PlaceholderList';

const TopTracksList = styled.ol`
  margin: 0;
  padding: 0 0 0.5rem 0;
  color: ${props => props.theme.colors.e};
  counter-reset: top-tracks-counter;
  list-style: none;
`;

export class TopTracks extends Component {
  static propTypes = {
    token: PropTypes.string,
    tracks: PropTypes.object,
    loading: PropTypes.bool,
    error: PropTypes.object,
    fetchTracks: PropTypes.func
  };

  componentDidMount() {
    const timeRange = localStorage.getItem(appTimeRange);

    if (timeRange) {
      this.props.fetchTracks(this.props.token, timeRange);
    } else {
      this.props.fetchTracks(this.props.token);
    }
  }

  render() {
    const { error, loading, tracks } = this.props;

    if (error) {
      return handleError(error);
    }

    if (loading) {
      return <PlaceholderList items={5} />;
    }

    return (
      <>
        <TopTracksList>
          {!isEmpty(tracks) &&
            tracks.items.map(item => {
              return (
                <Track
                  key={item.id}
                  songName={item.name}
                  artists={item.artists}
                  images={item.album.images}
                  uri={item.uri}
                />
              );
            })}
        </TopTracksList>
        <TopTracksActions />
      </>
    );
  }
}

const mapStateToProps = state => ({
  token: getToken(state),
  tracks: getTastesTracks(state),
  loading: tastesLoading(state),
  error: tastesError(state)
});

const mapDispatchToProps = dispatch => ({
  fetchTracks: (token, range) => dispatch(getTopTracks(token, range))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TopTracks);
