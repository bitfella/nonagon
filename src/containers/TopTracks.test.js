import React from 'react';
import { shallow } from 'enzyme';
import { TopTracks } from './TopTracks';

const props = {
  token: 'ACCESS_TOKEN',
  tracks: {},
  loading: false,
  error: {},
  fetchTracks: jest.fn()
};

describe('TopTracks', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<TopTracks {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
