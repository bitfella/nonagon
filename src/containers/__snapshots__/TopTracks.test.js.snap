// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopTracks renders correctly 1`] = `<Error />`;
